ts/private/ts_project_worker.js

```diff
@@ -279,32 +279,36 @@ function createEmitAndLibCacheAndDiagnosticsProgram(
     );
 
     /** Emit Cache */
-    /** @type {Map} */
-    const emittedFiles = (host.emittedFiles = host.emittedFiles || new Map());
-    /** @type {Map} */
-    const emittingMap = (host.emittedFilesWeak = host.emittedFilesWeak || new Map());
+    const NOT_FROM_SOURCE = Symbol.for("NOT_FROM_SOURCE")
+    /** @type {Map<string, string>} */
+    const outputSourceMapping = (host.outputSourceMapping = host.outputSourceMapping || new Map());
+    /** @type {Map<string, {text: string, writeByteOrderMark: boolean}>} */
+    const outputCache = (host.outputCache = host.outputCache || new Map());
 
     const emit = builder.emit;
     builder.emit = (targetSourceFile, writeFile, cancellationToken, emitOnlyDtsFiles, customTransformers) => {
         writeFile = writeFile || host.writeFile;
         if (!targetSourceFile) {
-            for (const [path, content] of emittedFiles.entries()) {
-                const sourcePath = emittingMap.get(path);
-                if (!builder.getSourceFile(sourcePath)) {
-                    emittedFiles.delete(path);
-                    emittingMap.delete(path);
-                } else {
-                    writeFile(path, content);
-                }
+            for (const [path, entry] of outputCache.entries()) {
+                const sourcePath = outputSourceMapping.get(path);
+                // if the source is not part of the program anymore, then drop the output from the output cache.
+                if (sourcePath != NOT_FROM_SOURCE && !builder.getSourceFile(sourcePath)) {
+                    outputSourceMapping.delete(path);
+                    outputCache.delete(path);
+                    continue;
+                } 
+                writeFile(path, entry.text, entry.writeByteOrderMark);
             }
         }
 
-        const writeF = (fileName, data, writeByteOrderMark, onError, sourceFiles) => {
-            writeFile(fileName, data, writeByteOrderMark, onError, sourceFiles);
-            if (sourceFiles && sourceFiles.length !== 0) {
-                host.debuglog?.(`putting ${fileName} into emit cache`);
-                emittingMap.set(fileName, sourceFiles[0].fileName);
-                emittedFiles.set(fileName, data);
+        const writeF = (fileName, text, writeByteOrderMark, onError, sourceFiles) => {
+            writeFile(fileName, text, writeByteOrderMark, onError, sourceFiles);
+            outputCache.set(fileName, {text, writeByteOrderMark});
+            if (sourceFiles?.length > 0) {
+                outputSourceMapping.set(fileName, sourceFiles[0].fileName)
+            } else {
+                // if the file write is not the result of a source mark it as not from source not avoid cache drops.
+                outputSourceMapping.set(fileName, NOT_FROM_SOURCE)
```

Comment:
I guess this is the fix for tsbuildinfo?

Reply:
OK, I understand the fix now. tsbuildinfo file doesn't have any source, so it's sourceFiles.length will always be zero and hence the bug.